x/eibc/keeper/demand_order.go

```diff
@@ -0,0 +1,186 @@
+		unprefixedDenom := fungibleTokenPacketData.Denom[len(voucherPrefix):]
+		// coin denomination used in sending from the escrow address
+		denom = unprefixedDenom
+		// The denomination used to send the coins is either the native denom or the hash of the path
+		// if the denomination is not native.
+		denomTrace := transfertypes.ParseDenomTrace(unprefixedDenom)
+		if denomTrace.Path != "" {
+			denom = denomTrace.IBCDenom()
+		}
+	} else {
+		denom = utils.GetForeignIBCDenom(packet.GetDestChannel(), fungibleTokenPacketData.Denom)
+	}
+	return denom
+}
+
+/*
+In case of timeout/errack:
+
+	fee = fee_multiplier*transfer_amount
+	price = transfer_amount-fee
+	order is created with (price,fee)
+	the order creator immediately receives price on fulfillment from fulfiller
+	when the ack/timeout finalizes, the fulfiller receives the transfer_amount
+
+	therefore:
+	fulfiller balance += (transfer_amount - (transfer_amount-fee))
+	equivalent to += fee_multiplier*transfer_amount
+	demander balance += (transfer_amount - fee)
+	equivalent to += (1-fee_multiplier)*transfer_amount
+*/
+func (k Keeper) PrepareProtocolDemandOrder(ctx sdk.Context, fungibleTokenPacketData transfertypes.FungibleTokenPacketData,
+	rollappPacket *commontypes.RollappPacket,
+) (*eibctypes.DemandOrder, error) {
+	// Validate the fungible token packet data as we're going to use it to create the demand order
+	if err := fungibleTokenPacketData.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	// Verify the original recipient is not a blocked sender otherwise could potentially use eibc to bypass it
+	if k.BlockedAddr(fungibleTokenPacketData.Receiver) {
+		return nil, fmt.Errorf("not allowed to receive funds: receiver: %s", fungibleTokenPacketData.Receiver)
+	}
+	// Calculate the demand order price and validate it,
+	amt, _ := sdk.NewIntFromString(fungibleTokenPacketData.Amount) // guaranteed ok and positive by above validation
+
+	// Calculate the fee by multiplying the fee by the price
+	var feeMultiplier sdk.Dec
+	switch rollappPacket.Type {
+	case commontypes.RollappPacket_ON_TIMEOUT:
+		feeMultiplier = k.TimeoutFee(ctx)
+	case commontypes.RollappPacket_ON_ACK:
+		feeMultiplier = k.ErrAckFee(ctx)
+	}
+	fee := feeMultiplier.MulInt(amt).TruncateInt()
+	if !fee.IsPositive() {
+		ctx.Logger().Debug("fee is not positive, skipping demand order creation", "packet", rollappPacket.LogString())
+		return nil, nil
+	}
+	demandOrderPrice := amt.Sub(fee)
+
+	trace := transfertypes.ParseDenomTrace(fungibleTokenPacketData.Denom)
+	demandOrderDenom := trace.IBCDenom()
+	demandOrderRecipient := fungibleTokenPacketData.Sender // and who tried to send it (refund because it failed)
+
+	order := eibctypes.NewDemandOrder(*rollappPacket, demandOrderPrice, fee, demandOrderDenom, demandOrderRecipient)
+	if err := order.Validate(); err != nil {
+		return nil, fmt.Errorf("validate eibc data: %w", err)
+	}
+	return order, nil
```

Comment:
duplicated, could be dried out

Reply:
dried out some
but the creation I prefer to do in the inner func, otherwise I need to return all args (`demandOrderPrice, fee, demandOrderDenom, demandOrderRecipient`)